src/response.rs

```diff
@@ -0,0 +1,206 @@
+use serde::de::{self, Deserializer, MapAccess, Visitor};
+use serde::Deserialize;
+use std::fmt;
+use std::str::FromStr;
+use std::time::SystemTime;
+
+#[derive(Debug, PartialEq)]
+pub struct PeriodTime(SystemTime);
+
+impl From<SystemTime> for PeriodTime {
+    fn from(st: SystemTime) -> Self {
+        PeriodTime(st)
+    }
+}
+
+impl From<PeriodTime> for SystemTime {
+    fn from(pt: PeriodTime) -> Self {
+        pt.0
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename = "usage_report")]
+pub struct UsageReport {
+    pub metric: String,
+    pub period: Period,
+    pub period_start: PeriodTime,
+    pub period_end: PeriodTime,
+    pub max_value: u64,
+    pub current_value: u64,
+}
+
+// Unfortunately the XML output from Apisonator includes a rather useless "usage_reports" tag that
+// is then followed by a "usage_report" tag in each UsageReport, so we need to wrap that up.
+#[derive(Debug, Deserialize, PartialEq)]
+pub enum UsageReports {
+    #[serde(rename = "usage_report")]
+    UsageReports(Vec<UsageReport>),
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename = "status")]
+pub struct Authorization {
+    authorized: bool,
+    plan: String,
+    usage_reports: UsageReports,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct PeriodInstance {
+    start: SystemTime,
+    end: SystemTime,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Period {
+    Minute,
+    Hour,
+    Day,
+    Week,
+    Month,
+    Year,
+    Eternity,
+}
+
+struct PeriodStringVisitor;
+
+impl<'de> Visitor<'de> for PeriodStringVisitor {
+    type Value = Period;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a string that represents a period")
+    }
+
+    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        match v {
+            "minute" => Ok(Period::Minute),
+            "hour" => Ok(Period::Hour),
+            "day" => Ok(Period::Day),
+            "week" => Ok(Period::Week),
+            "month" => Ok(Period::Month),
+            "year" => Ok(Period::Year),
+            "eternity" => Ok(Period::Eternity),
+            _ => Err(E::custom("Invalid period")),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Period {
+    fn deserialize<D>(deserializer: D) -> Result<Period, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_any(PeriodStringVisitor)
+    }
+}
+
+struct TimestampVisitor;
+
+impl<'de> Visitor<'de> for TimestampVisitor {
+    type Value = PeriodTime;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a string that represents a timestamp")
+    }
+
+    fn visit_map<V>(self, mut map: V) -> Result<PeriodTime, V::Error>
+    where
+        V: MapAccess<'de>,
+    {
+        // We know there's only one key with one value - Note: probably safe to skip
+        let _key: Option<String> = map.next_key()?;
```

Comment:
this was triggering an warning because it is unused - made it silent for now, but tested just removing it and it seemed to work, maybe with some more confirmation we can remove it

Reply:
`next_key` needs to be there, but I forgot to add a comment with the reason: https://github.com/serde-rs/serde/blob/535e3d4372f8de923dd5c0042d425ee3b13a389b/serde/src/de/mod.rs#L1833